### PR TITLE
Syntax highlighting : first update (added several types of highlights…

### PR DESCRIPTION
**Pull request by francois-lefoulon**
… as in emacs).

(I accidentally added an emacs modification, it should be ignored)

### CODE REVIEW

**Comment by fblanqui** [comment]
Thanks @francois-lefoulon for your PR. Thanks to fix the modif in the emacs mode. @elhaddadyacine uses VSCode. He could give some useful feedbacks.

File: editors/emacs/lambdapi-input.el [hunk 1]
```diff
@@ -35,7 +35,7 @@
 
  ;; Miscellaneous
 
- ("\\Box" ?□)
+ ("\\Box" ?□)
```
Comment (fblanqui) [inline editors/emacs/lambdapi-input.el:38]: Apparemment, il reste un espace en trop... ;-)

File: editors/vscode/.vscode/settings.json [hunk 2]
```diff
@@ -7,5 +7,6 @@
         "out": true // set this to false to include "out" folder in search results
     },
     // Turn off tsc task auto detection since we have the necessary tasks as npm scripts
-    "typescript.tsc.autoDetect": "off"
+    "typescript.tsc.autoDetect": "off",
+
```
Comment (fblanqui) [inline editors/vscode/.vscode/settings.json:11]: Is this modification really useful? (It's better to minimize diffs)

File: editors/vscode/package.json [hunk 3]
```diff
@@ -62,6 +62,7 @@
         "category": "Lambdapi"
       }
     ]
+
```
Comment (fblanqui) [inline editors/vscode/package.json:65]: Idem.

**Comment by francois-lefoulon** [comment]
As a reminder : 
ctrl+alt+c : Toggle navigation with keybindings
ctrl+alt+s : Go forward
ctrl+alt+z : Go backwards

**Comment by amelieled** [comment]
Hi @francois-lefoulon !
In CoqIde:
Ctrl + down arrow: Go forward
Ctrl + up arrow: Go backwards
Ctrl + ->: Check the file until the cursor
Ctrl + "Fin"/origin: Check the file

and the user can change these keybindings!
To be honest with you, the CoqIde's keybindings are very practical and the using of buttons isn't really efficient, but as you want! :)

**Comment by francois-lefoulon** [comment]
Hello @amelieled ! 
Thanks for sharing, I checked out VS Coq but not CoqIDE; in VS COQ it's basically the same but with alt+arrow !
I think I'll provide the keybindings with arrows as in CoqIDE for better UX, even though I prefer using z/s/q/d as arrows personally.
Anyway, there are no "buttons" in VS Code, you can just search for a command in a search bar so keybindings are... key x) I don't know if you can change keybindings though.

**Comment by fblanqui** [comment]
Hello @francois-lefoulon . Thanks @amelieled for your remarks. I think that it would good to have as few key strokes as possible because they will have to be done very often. So I prefer Ctrl-downarrow to Ctrl-Alt-downarrow, which is used for navigating between workspaces in some systems. Or is Ctrl-downarrow often map to something else already in VSCode?

**Comment by francois-lefoulon** [comment]
> Hello @francois-lefoulon . Thanks @amelieled for your remarks. I think that it would good to have as few key strokes as possible because they will have to be done very often. So I prefer Ctrl-downarrow to Ctrl-Alt-downarrow, which is used for navigating between workspaces in some systems. Or is Ctrl-downarrow often map to something else already in VSCode?

Don't worry, I planned on "Ctrl+arrow", the "Ctrl+Alt+..." is due to "alt+s" and "ctrl+s" being already in use. I didn't add the other shortcuts yet though, I was working on latex snippets.

Is there still anything preventing my commits from being merged ?

**Comment by francois-lefoulon** [comment]
About my last commit, for instance typing @a and pressing tab or enter will type an alpha.

**Comment by fblanqui** [comment]
Hi @francois-lefoulon . Goals are currently printed surrounded by accolades. Could you please remove them?

**Comment by fblanqui** [comment]
@francois-lefoulon you can resolve a conversation when you fixed the problem raised.

Concerning unicode symbols, it's not sure it is a good idea to use `@` as it has a specific meaning in the syntax of LP. Why not using backquote like in emacs?

**Comment by fblanqui** [comment]
But why after all defining our own key-bindings? We could perhaps try to reuse existing ones. In particular, it would be nice to find a plugin mapping Latex commands to unicode caracters (e.g. \alpha to α). I found some plugins for entering unicode caracters in vscode that we could advice in the doc:

https://marketplace.visualstudio.com/items?itemName=oijaz.unicode-latex

https://marketplace.visualstudio.com/items?itemName=GuidoTapia2.unicode-math-vscode
table: https://github.com/mvoidex/UnicodeMath/blob/master/table.md

https://marketplace.visualstudio.com/items?itemName=brunnerh.insert-unicode

Other links are welcome.

**Comment by fblanqui** [comment]
@francois-lefoulon Thanks to add a file doc/vscode.md where you will give details about the VSCode interface (overview, features, key-bindings for proof navigation).

**Comment by fblanqui** [comment]
https://marketplace.visualstudio.com/items?itemName=oijaz.unicode-latex works very well.

On the other had, I cannot make https://marketplace.visualstudio.com/items?itemName=GuidoTapia2.unicode-math-vscode work. Do you know how it works? This is unfortunate because it already defines key kindings for ->, --> and == for instance.

If we cannot make it work, we could add our own key bindings for these symbols.

**Comment by fblanqui** [comment]
When I do Ctrl-Alt-c, I get an error: `/!\ command 'extension.cscode-lp.cursor' not found.`.

**Comment by francois-lefoulon** [comment]
> Hi @francois-lefoulon . Goals are currently printed surrounded by accolades. Could you please remove them?

Of course.



> @francois-lefoulon you can resolve a conversation when you fixed the problem raised.
> 
> Concerning unicode symbols, it's not sure it is a good idea to use `@` as it has a specific meaning in the syntax of LP. Why not using backquote like in emacs?

Typing "@" does not force you to add a latex character, since you have to press enter or tab after typing the snippet. I think backtick is an extremely unpractical shortcut, when you have your right fingers on jklm and don't necessarily want to look at your keyboard, but it is subjective.



> But why after all defining our own key-bindings? We could perhaps try to reuse existing ones. In particular, it would be nice to find a plugin mapping Latex commands to unicode caracters (e.g. \alpha to α). I found some plugins for entering unicode caracters in vscode that we could advice in the doc:
> 
> https://marketplace.visualstudio.com/items?itemName=oijaz.unicode-latex
> 
> https://marketplace.visualstudio.com/items?itemName=GuidoTapia2.unicode-math-vscode
> table: https://github.com/mvoidex/UnicodeMath/blob/master/table.md
> 
> https://marketplace.visualstudio.com/items?itemName=brunnerh.insert-unicode
> 
> Other links are welcome.

I was actually inspired by 
https://github.com/James-Yu/LaTeX-Workshop
which uses "@+letter" but I'll definitely check out the other links.



> @francois-lefoulon Thanks to add a file doc/vscode.md where you will give details about the VSCode interface (overview, features, key-bindings for proof navigation).

Alright.



> https://marketplace.visualstudio.com/items?itemName=oijaz.unicode-latex works very well.
> 
> On the other had, I cannot make https://marketplace.visualstudio.com/items?itemName=GuidoTapia2.unicode-math-vscode work. Do you know how it works? This is unfortunate because it already defines key kindings for ->, --> and == for instance.
> 
> If we cannot make it work, we could add our own key bindings for these symbols.

Now that I have defined keybindings in our own extension, it is easier to add new ones than to merge with another extension. Any suggestions for new snippets is welcome, and I still didn't add all the ones present in emacs.



> When I do Ctrl-Alt-c, I get an error: `/!\ command 'extension.cscode-lp.cursor' not found.`.

I'll come back to you as soon as I find the issue.

**Comment by francois-lefoulon** [comment]
> When I do Ctrl-Alt-c, I get an error: `/!\ command 'extension.cscode-lp.cursor' not found.`.

is it "cscode" or "vscode" ? I don't see any "extension.cscode-lp.cursor" on git nor on my local files (the right command is "extension.vscode-lp.cursor" present in "package.json", l.75 and 93, and in "client.ts", l.77). Have you accidentally modified the code ?

It works fine on my part, maybe there is a dependency with another file I've modified since last commit, but it's unlikely.

**Comment by fblanqui** [comment]
It is vscode.

**Comment by francois-lefoulon** [comment]
> It is vscode.

Does it still print the same error since I pushed ? Have you tried to make clean and make, and closing VS Code afterwards ? (For instance I still had brackets around my Goals after make clean and make, but then I closed and reopened the editor and it worked fine)

File: doc/vscode.md [hunk 4]
```diff
@@ -18,3 +18,47 @@ sudo apt install npm node-typescript
...
+
+Goals are visualised in a panel on the right side of the editor.
+
+- By default, goals are printed according to the current position of the cursor.
+
+- By toggling off "cursor mode" with
+```
+Ctrl+Alt+C
+```
+$\hspace{0.8cm}$ you can navigate the proof with [keybindings](#keybindings).
```
Comment (fblanqui) [inline doc/vscode.md:41]: `$\hspace{0.8cm}$` is not MD.

File: doc/vscode.md [hunk 5]
```diff
@@ -18,3 +18,47 @@ sudo apt install npm node-typescript
...
+- Navigate down : go forward in the proof
+
+```
+Ctrl+Alt+S
+```
+
+- Navigate up : go backwards in the proof
+
+```
+Ctrl+Alt+Z
```
Comment (fblanqui) [inline doc/vscode.md:63]: Please change to arrows.

**Comment by fblanqui** [comment]
Yes, I still have the problem. I use vscode 1.46.1.

**Comment by fblanqui** [comment]
For the record, link to the new doc: https://github.com/francois-lefoulon/lambdapi/blob/contribution/doc/vscode.md

**Comment by fblanqui** [comment]
Sorry, I forgot to recompile... I do not have the error anymore.

**Comment by fblanqui** [comment]
But proof navigation does seem to work...

**Comment by francois-lefoulon** [comment]
I enabled keybindings navigation by default as a workaround, you don't need ctrl+alt+c anymore

**Comment by fblanqui** [comment]
Ctrl-up/down doesn't work: no goal are printed. And what do you do to visualize the "proof-cursor"? See https://github.com/Deducteam/lambdapi/pull/414#issuecomment-647459820 .

**Comment by fblanqui** [comment]
By the way, what is the directory editors/vscode/.vscode for? Is it normal to have this directory under git or is it a mistake? @ejgallego , @francois-lefoulon?

**Comment by fblanqui** [comment]
@francois-lefoulon you can mark a conversation as resolved when you fully fixed a problem.

**Comment by francois-lefoulon** [comment]
> By the way, what is the directory editors/vscode/.vscode for? Is it normal to have this directory under git or is it a mistake? @ejgallego , @francois-lefoulon? Link: https://github.com/Deducteam/lambdapi/tree/master/editors/vscode/.vscode

A .vscode is generated whenever you create a VS Code extension to set up some parameters while you work in the extension folder as working directory, I think. So it's not a mistake.



> @francois-lefoulon you can mark a conversation as resolved when you fully fixed a problem.

I know, I was just waiting for your answers, especially when it does not seem to work for you.





> Ctrl-up/down doesn't work: no goal are printed. And what do you do to visualize the "proof-cursor"? See [#414 (comment)](https://github.com/Deducteam/lambdapi/pull/414#issuecomment-647459820) .

The current line for proof navigation is highlighted in green.

image

It works fine on my part, and we have the same version for VS Code, so I really don't understand what's wrong. Try to open the extension folder in vscode ("code ." while in editors/vscode) and press F5.

**Comment by fblanqui** [comment]
When I launch `code .` in editors/vscode/, I am asked: "This workspace has extension recommendations. Install All? Show recommendations?" Recommendations are TSLint 1.2.3. Should I install it?

**Comment by francois-lefoulon** [comment]
> When I launch `code .` in editors/vscode/, I am asked: "This workspace has extension recommendations. Install All? Show recommendations?" Recommendations are TSLint 1.2.3. Should I install it?

Well, I have TS Lint, so it can't hurt. I it works for you then I'll have to investigate why there is a dependency here.

**Comment by francois-lefoulon** [comment]
TS Lint is just support for TypeScript verification though, I don't know what it has to do with the LP extension.

**Comment by fblanqui** [comment]
Well it is just a recommendation and can probably be ignored.
Anyway, if I do `code .` in editors/vscode/, then open a file in a different folder, for instance ../../mylib/Nat.lp, then no proof is displayed...

**Comment by fblanqui** [comment]
Well, currently, I don't have proof display anymore, even if I open the file directly from the folder where it is!...

**Comment by fblanqui** [comment]
Ah I need to do Ctrl-Alt-c to display the goals.

**Comment by francois-lefoulon** [comment]
This doesn't make sense, I changed the default... It's as if your VS Code is stuck in a previous version. Maybe try to uninstall it in the extensions panel (last icon in the left side bar), and then reinstalling it with make clean and make

**Comment by fblanqui** [comment]
Ah I see. Now, when I do Ctrl-arrows, I indeed get a command in a box with green background and can move it. What is a little bit surprising is that you have nothing at the beginning when you open a file. Then, if you click somewhere in a proof, nothing is displayed. This is disturbing.

**Comment by fblanqui** [comment]
Perhaps you should highlight the first command by default. Would it be possible?

**Comment by francois-lefoulon** [comment]
> Perhaps you should highlight the first command by default. Would it be possible?

Yes. By the way, the highlight is always one line late, I need to fix this.

**Comment by fblanqui** [comment]
Also I think that it should be better to do like in PG: do not put a command in a box but just change the color background of all the buffer until the command where is the proof-cursor.

**Comment by francois-lefoulon** [comment]
> Also I think that it should be better to do like in PG: do not put a command in a box but just change the color background of all the buffer until the command where is the proof-cursor.

VS Coq does the same thing. I thought it was distracting to have most of your code highlighted in green, but I'm not the one making the proofs.

**Comment by fblanqui** [comment]
The background color should not be too flashy.
If this is not so many changes, why not doing it in this PR?

**Comment by fblanqui** [comment]
Thanks @francois-lefoulon . Would it be possible to remove the white frames? And, when you start code, to highlight the first command?

**Comment by francois-lefoulon** [comment]
> Thanks @francois-lefoulon . Would it be possible to remove the white frames? And, when you start code, to highlight the first command?

The frames are now removed. I had already highlighted the first command in f503089.

File: doc/vscode.md [hunk 6]
```diff
@@ -18,3 +18,43 @@ sudo apt install npm node-typescript
...
+
+# Features
+
+
+##  Proof navigation
+
+Goals are visualised in a panel on the right side of the editor.
+
+- You can navigate the proof with [keybindings](#keybindings).
+-
```
Comment (fblanqui) [inline doc/vscode.md:36]: empty bullet

File: doc/vscode.md [hunk 7]
```diff
@@ -18,3 +18,43 @@ sudo apt install npm node-typescript
...
+# Features
+
+
+##  Proof navigation
+
+Goals are visualised in a panel on the right side of the editor.
+
+- You can navigate the proof with [keybindings](#keybindings).
+- 
+- "Cursor mode" can be toggled to navigate the proff with the cursor.
```
Comment (fblanqui) [inline doc/vscode.md:37]: typo: proof

**Comment by fblanqui** [comment]
A few functionalities are missing for merging this PR:
- shortcuts for inputing some unicode caracters used in the syntax (arrows) like in the emacs mode
- Ctrl-Enter to move the proof-cursor to where is the writing-cursor

File: doc/vscode.md [hunk 8]
```diff
@@ -18,3 +18,43 @@ sudo apt install npm node-typescript
...
+
+# Keybindings
+
+- Navigate down : go forward in the proof
+
+```
+Ctrl+Down
+```
+
+- Navigate up : go backwards in the proof
```
Comment (fblanqui) [inline doc/vscode.md:51]: typo: backward

File: doc/vscode.md [hunk 9]
```diff
@@ -18,3 +18,43 @@ sudo apt install npm node-typescript
...
+- 
+- "Cursor mode" can be toggled to navigate the proff with the cursor.
+
+##  Snippets
+
+Typing "@letter" will suggest a corresponding unicode greek letter (for instance "@b" will suggest β). Press Enter or Tab to confirm adding the Unicode character.
+
+# Keybindings
+
+- Navigate down : go forward in the proof
```
Comment (fblanqui) [inline doc/vscode.md:45]: Perhaps it is better to present things the other way around as follows:
key combination: action
```
- Ctrl-Down: go one step forward
- Ctrl-Up: go one step backward
- Ctrl-Enter: go to where is the cursor
- Ctrl-Alt-c: toggle cursor mode (follow the cursor or not)
```
By the way, is it Ctrl-Alt-c or Ctrl-Alt-C?

**Comment by fblanqui** [comment]
For the record, the doc is on https://github.com/francois-lefoulon/lambdapi/blob/contribution/doc/vscode.md

File: doc/vscode.md [hunk 10]
```diff
@@ -30,31 +30,78 @@ It is still under development, thus might be incomplete and buggy.
...
+
+- ```@eq```: ≡
+
+- ```@as```: ⊢
+
+- ```@fa```: ∀
+
+- ```@ex```: ∃
+
+- ```@imp```: ⇒
```
Comment (fblanqui) [inline doc/vscode.md:92]: remove (done twice)

File: doc/vscode.md [hunk 11]
```diff
@@ -30,31 +30,78 @@ It is still under development, thus might be incomplete and buggy.
...
+- ```@itletter```(@itn): italic letter (𝑁)
+
+- ```@ibletter```(@bin): italic bold (𝑵)
+
+- ```@scletter```(@scn): script letter (𝒩 )
+
+- ```@bsletter```(@bsn): bold script letter (𝓝)
+
+- ```@frletter```(@frn): Fraktur letter (𝔑)
+
```
Comment (fblanqui) [inline doc/vscode.md:65]: I propose to simplify some prefixes: d for double-struck, i for italic, s for script, bs for bold script, f for fraktur. Isn't there a bold (without italic) too?

File: doc/vscode.md [hunk 12]
```diff
@@ -30,31 +30,78 @@ It is still under development, thus might be incomplete and buggy.
...
+- ```@and``` : ∧
+
+- ```@neg``` : ¬
+
+- ```@imp```: ⇒
+
+- ```@ar```: →
+
+- ```@ho```: ↪
```
Comment (fblanqui) [inline doc/vscode.md:80]: I would prefer `@re`. Isn't it possible to have `->` replaced by ↪ too?

File: doc/vscode.md [hunk 13]
```diff
@@ -30,31 +30,78 @@ It is still under development, thus might be incomplete and buggy.
...
+
+- ```@or``` : ∨
+
+- ```@and``` : ∧
+
+- ```@neg``` : ¬
+
+- ```@imp```: ⇒
+
+- ```@ar```: →
```
Comment (fblanqui) [inline doc/vscode.md:78]: I would prefer `@ra`. Isn't it possible to have `->` replaced by →?

File: doc/vscode.md [hunk 14]
```diff
@@ -30,31 +30,78 @@ It is still under development, thus might be incomplete and buggy.
...
+
+---
+
+###   Common symbols
+
+- ```@or``` : ∨
+
+- ```@and``` : ∧
+
+- ```@neg``` : ¬
```
Comment (fblanqui) [inline doc/vscode.md:74]: `@not`

File: doc/vscode.md [hunk 15]
```diff
@@ -30,31 +30,78 @@ It is still under development, thus might be incomplete and buggy.
...
+
+- ```@imp```: ⇒
+
+- ```@ar```: →
+
+- ```@ho```: ↪
+
+- ```@set```: ≔
```
Comment (fblanqui) [inline doc/vscode.md:82]: Isn't it possible to have `:=` replaced by ≔?

File: doc/vscode.md [hunk 16]
```diff
@@ -30,31 +30,78 @@ It is still under development, thus might be incomplete and buggy.
...
+
+- ```@eq```: ≡
+
+- ```@as```: ⊢
```
Comment (fblanqui) [inline doc/vscode.md:86]: `@th` or, better, `|-`.

File: doc/vscode.md [hunk 17]
```diff
@@ -30,31 +30,78 @@ It is still under development, thus might be incomplete and buggy.
...
+
+- ```@fa```: ∀
```
Comment (fblanqui) [inline doc/vscode.md:88]: `@all`

File: doc/vscode.md [hunk 18]
```diff
@@ -30,31 +30,78 @@ It is still under development, thus might be incomplete and buggy.
...
+- ```@imp```: ⇒
+
+- ```@ar```: →
+
+- ```@ho```: ↪
+
+- ```@set```: ≔
+
+- ```@eq```: ≡
+
```
Comment (fblanqui) [inline doc/vscode.md:85]: `==` ?

File: doc/vscode.md [hunk 19]
```diff
@@ -18,3 +18,113 @@ sudo apt install npm node-typescript
...
+- ```"`not"``` : ¬
+
+- ```"`imp"```: ⇒
+
+- ```"`ra"```: →
+
+- ```"`re"```: ↪
+
+- ```"`set"```: ≔
```
Comment (fblanqui) [inline doc/vscode.md:84]: I would prefer `is`.

File: doc/vscode.md [hunk 20]
```diff
@@ -18,3 +18,113 @@ sudo apt install npm node-typescript
...
+
+- ```"`eq"```: ≡
+
+- ```"`th"```: ⊢
+
+- ```"`all"```: ∀
+
+- ```"`ex"```: ∃
+
+- ```"`imp"```: ⇒
```
Comment (fblanqui) [inline doc/vscode.md:94]: On github, I see it already defined on line 78...

File: doc/vscode.md [hunk 21]
```diff
@@ -18,3 +18,113 @@ sudo apt install npm node-typescript
...
+
+- ```"`box"```: □
+
+- ```"`tf"```: ⸬
```
Comment (fblanqui) [inline doc/vscode.md:98]: What is "tf" for?

Comment (fblanqui) [reply doc/vscode.md:98]: I would prefer "`cons".

File: doc/vscode.md [hunk 22]
```diff
@@ -18,3 +18,113 @@ sudo apt install npm node-typescript
...
+If snippet completion does not seem to work with the VS-Code Lambdapi Extension or the recommended extension, try pressing "Ctrl+Space" to see completion suggestions.
+
+# Keybindings
+
+For proof navigation :
+
+- ```Ctrl+Down``` : go one step forward
+- ```Ctrl+Up``` : go one step backward
+- `Ctrl+Left` : go to the previous proof (or the beginning)
+- `Ctrl+Right` : go to the next proof (or the end)
```
Comment (fblanqui) [inline doc/vscode.md:119]: François, could you please use Ctrl+down/up for next/previous proof, and Ctrl+right/left for next/previous step?

File: doc/vscode.md [hunk 23]
```diff
@@ -18,3 +18,113 @@ sudo apt install npm node-typescript
...
+
+---
+
+###   Fonts
+
+- ```"`dletter"```(for instance `dn): double-struck letter (ℕ)
+
+- ```"`iletter"```(`in): italic letter (𝑁)
+
+- ```"`ibletter"```(`ibn): italic bold (𝑵)
```
Comment (fblanqui) [inline doc/vscode.md:60]: ib -> I (like in emacs)

File: doc/vscode.md [hunk 24]
```diff
@@ -18,3 +18,113 @@ sudo apt install npm node-typescript
...
+
+- ```"`sletter"```(`sn): script letter (𝒩 )
+
+- ```"`bsletter"```(`bsn): bold script letter (𝓝)
```
Comment (fblanqui) [inline doc/vscode.md:64]: bs -> S (like in emacs)

File: doc/vscode.md (hunk 21 above)
Comment (francois-lefoulon) [reply doc/vscode.md:98]: "therefore"

**Comment by fblanqui** [comment]
Thanks @francois-lefoulon to update your branch so that I can merge.

**Comment by francois-lefoulon** [comment]
Have I done this right ? I'm not used to contributing via pull requests. Can you merge now ?

**Comment by fblanqui** [comment]
Thanks @francois-lefoulon !

**Pull request merged**